cogs/music.py

```diff
@@ -0,0 +1,584 @@
+        ctx.voice_client.stop()
+        await ctx.send(
+            f"moved to `{self.skip_opts[0]}` in currently playing track.")
+
+    wrapper.__name__ = func.__name__
+    wrapper.__doc__ = func.__doc__
+
+    return wrapper
+
+
+class Music(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+        self.song_queue: deque = deque()
+        self.backup: deque = deque()
+        self.loop_queue: bool = False
+        self.song_lock: asyncio.Lock = asyncio.Lock()
+        self.playing = None
+        self.looping = None
+        self.volume_level: int = 100
+        self.skip_opts: Optional[tuple[str, int]] = None
+        self.song_start_time: float = 0
+
+    async def get_info(self, url: str):
+        return await self.bot.loop.run_in_executor(
+            None, partial(YTDL.extract_info, url, download=False))
+
+    def check_reaction(self, embed_msg: discord.Message,
+                       reaction: discord.Reaction, user: discord.User):
+        return (user != self.bot.user and reaction.message.id == embed_msg.id
+                and str(reaction) in QUEUE_ACTIONS)
+
+    @commands.command()
+    async def play(self, ctx, *, url: str = None):
+        """Streams from a youtube url or track name, or if none is given, from the queue"""
+
+        in_main: bool = ctx.voice_client is None
+
+        if ctx.author.voice is None:
+            await ctx.send(
+                "you are not currently connected to a voice channel.")
+            return
+
+        if in_main and (not self.song_queue) and url is None:
+            await ctx.send("you did not specify a song to play.")
+            return
+
+        if not in_main:
+            if ctx.voice_client.is_paused():
+                await ctx.send(
+                    "bot is currently paused, please use `resume` before playing."
+                )
+                return
+            if ctx.author.voice.channel != ctx.voice_client.channel:
+                if len(ctx.voice_client.channel.members) > 1:
+                    await ctx.send(
+                        "bot is currently playing music for users in another voice channel."
+                    )
+                    return
+                ctx.voice_client.pause()
+                await ctx.voice_client.move_to(ctx.author.voice.channel)
+                if not url:
+                    ctx.voice_client.resume()
+            elif url is None:
+                await ctx.send(
+                    "bot is currently playing a song and you did not specify a new song to play."
+                )
+                return
+
+        def release_lock(_):
+            self.song_lock.release()
+
+        if url:
+            await ctx.trigger_typing()
+            data = await self.get_info(url)
+            entries = data.get("entries", [data])
+            for track in reversed(entries):
+                self.song_queue.insert(0, (track, str(ctx.author)))
```

Comment:
deque's `extendleft` method might be able to make this a "one-liner" with an iterator - plus, entries is only used in one place here